dmoj/cptbox/filesystem_policies.py

```diff
@@ -24,25 +24,58 @@ class File:
     pass
 
 
-class ExactFile:
+class BaseFilesystemAccessRule:
+    repr_name = 'BaseFilesystemAccessRule'
+
     def __init__(self, path: str):
+        path = os.path.expanduser(path)
+        assert os.path.abspath(path) == path, 'FilesystemAccessRule must specify a normalized, absolute path to rule'
         self.path = path
+        self._assert_rule_type()
 
+    def _assert_rule_type(self) -> None:
+        if self.exists():
+            self._assert_dir_type(os.path.isdir(self.path))
 
-class ExactDir:
-    access_mode = AccessMode.EXACT
+    def _assert_dir_type(self, is_dir: bool) -> None:
+        raise NotImplementedError
 
-    def __init__(self, path: str):
-        self.path = path
+    def exists(self):
+        return os.path.exists(self.path)
 
+    def realpath(self):
+        return os.path.realpath(self.path)
 
-class RecursiveDir:
-    access_mode = AccessMode.RECURSIVE
+    def is_realpath(self) -> bool:
+        return self.path.startswith('/proc/self') or self.realpath() == self.path  # proc/self should not be resolved
 
-    def __init__(self, path: str):
-        self.path = path
+    def __repr__(self) -> str:
+        return f'<{self.repr_name} {self.path}>'
```

Comment:
Why are you using this `repr_name` instead of just `__class__.__name__` or something?

Reply:
Wasn't aware of its existence. Pushed a new update to use it, thanks!